certs/service.go

```diff
@@ -0,0 +1,303 @@
+	errFailedCertRevocation      = errors.New("failed to revoke certificate")
+)
+
+var _ Service = (*certsService)(nil)
+
+// Service specifies an API that must be fulfilled by the domain service
+// implementation, and all of its decorators (e.g. logging & metrics).
+type Service interface {
+	// IssueCert issues certificate for given thing id if access is granted with token
+	IssueCert(ctx context.Context, token, thingID, daysValid string, keyBits int, keyType string) (Cert, error)
+
+	// ListCerts lists all certificates issued for given owner
+	ListCerts(ctx context.Context, token string, offset, limit uint64) (Page, error)
+
+	// RevokeCert revokes certificate for given thing
+	RevokeCert(ctx context.Context, token, thingID string) (Revoke, error)
+}
+
+type Config struct {
+	LogLevel       string
+	ClientTLS      bool
+	CaCerts        string
+	HTTPPort       string
+	ServerCert     string
+	ServerKey      string
+	BaseURL        string
+	ThingsPrefix   string
+	JaegerURL      string
+	AuthnURL       string
+	AuthnTimeout   time.Duration
+	SignTLSCert    tls.Certificate
+	SignX509Cert   *x509.Certificate
+	SignRSABits    int
+	SignHoursValid string
+	PKIHost        string
+	PKIPath        string
+	PKIRole        string
+	PKIToken       string
+}
+
+type certsService struct {
+	auth      mainflux.AuthNServiceClient
+	certsRepo Repository
+	sdk       mfsdk.SDK
+	conf      Config
+	pki       pki.Agent
+}
+
+// New returns new Certs service.
+func New(auth mainflux.AuthNServiceClient, certs Repository, sdk mfsdk.SDK, config Config, pki pki.Agent) Service {
+	return &certsService{
+		certsRepo: certs,
+		sdk:       sdk,
+		auth:      auth,
+		conf:      config,
+		pki:       pki,
+	}
+}
+
+type Revoke struct {
+	RevocationTime time.Time `mapstructure:"revocation_time"`
+}
+
+type Cert struct {
+	OwnerID        string    `json:"owner_id" mapstructure:"owner_id"`
+	ThingID        string    `json:"thing_id" mapstructure:"thing_id"`
+	ClientCert     string    `json:"client_cert" mapstructure:"certificate"`
+	IssuingCA      string    `json:"issuing_ca" mapstructure:"issuing_ca"`
+	CAChain        []string  `json:"ca_chain" mapstructure:"ca_chain"`
+	ClientKey      string    `json:"client_key" mapstructure:"private_key"`
+	PrivateKeyType string    `json:"private_key_type" mapstructure:"private_key_type"`
+	Serial         string    `json:"serial" mapstructure:"serial_number"`
+	Expire         time.Time `json:"expire" mapstructure:"-"`
+}
+
+func (cs *certsService) IssueCert(ctx context.Context, token, thingID string, daysValid string, keyBits int, keyType string) (Cert, error) {
+	var c Cert
+	owner, err := cs.auth.Identify(ctx, &mainflux.Token{Value: token})
+	if err != nil {
+		return c, errors.Wrap(ErrUnauthorizedAccess, err)
+	}
+
+	thing, err := cs.sdk.Thing(thingID, token)
+	if err != nil {
+		return c, errors.Wrap(errFailedCertCreation, err)
+	}
+
+	// If PKIClient == nil we don't use 3rd party PKI service.
```

Comment:
Please don't use code annotations for comments. btw, why you check `PKIHost` but the comment says `PKIClient`?